core/snapshot/engine.go

```diff
@@ -24,226 +22,165 @@ import (
+	// before the state as been reloaded in all providers.
+	preRestoreProviders []types.PreRestore
+	// postRestoreProviders tracks all the components that need to be called
+	// after the state as been reloaded in all providers.
+	postRestoreProviders []types.PostRestore
+
+	// snapshotTree hold the snapshot as an AVL tree.
+	snapshotTree *tree.Tree
+	// snapshotTreeLock is used every time it is needed to read from or writing to
+	// the AVL tree.
+	snapshotTreeLock sync.Mutex
+
+	// intervalBetweenSnapshots defines the internal between snapshots. The unit
+	// is based on the network commits. An interval of 10 means a snapshot is taken
+	// every 10 commits.
+	intervalBetweenSnapshots uint64
+	// commitsLeftBeforeSnapshot tracks the number of commits left before the
+	// engine snapshots the state of the node.
+	commitsLeftBeforeSnapshot uint64
+}
+
+// NewEngine returns a new snapshot engine.
+func NewEngine(ctx context.Context, vegaPath paths.Paths, conf Config, log *logging.Logger, timeSvc TimeService, statsSvc StatsService) (*Engine, error) {
 	if err := conf.Validate(); err != nil {
 		return nil, fmt.Errorf("invalid configuration: %w", err)
 	}
 
 	log = log.Named(namedLogger)
 	log.SetLevel(conf.Level.Get())
 
-	sctx, cfunc := context.WithCancel(ctx)
-	appPL := &types.PayloadAppState{
-		AppState: &types.AppState{},
-	}
-	app := appPL.Namespace()
+	appStatePayload := &types.PayloadAppState{}
+
 	eng := &Engine{
-		Config:       conf,
-		log:          log,
-		ctx:          sctx,
-		cfunc:        cfunc,
-		timeService:  tm,
-		statsService: stats,
-		namespaces:   []types.SnapshotNamespace{},
+		Config: conf,
+		log:    log,
+
+		timeService:  timeSvc,
+		statsService: statsSvc,
+
+		namespaces: []types.SnapshotNamespace{},
 		nsKeys: map[types.SnapshotNamespace][]string{
-			app: {appPL.Key()},
+			types.AppSnapshot: {appStatePayload.Key()},
 		},
-		nsTreeKeys: map[types.SnapshotNamespace][][]byte{
-			app: {
-				[]byte(types.KeyFromPayload(appPL)),
+		namespacesToTreeKeys: map[types.SnapshotNamespace][][]byte{
+			types.AppSnapshot: {
+				[]byte(types.KeyFromPayload(appStatePayload)),
 			},
 		},
 		treeKeyProvider: map[string]string{},
-		providers:       map[string]types.StateProvider{},
-		providersNS:     map[types.SnapshotNamespace][]types.StateProvider{},
-		versions:        make([]int64, 0, conf.KeepRecent), // cap determines how many versions we keep
-		versionHeight:   map[uint64]int64{},
-		wrap:            appPL,
-		app:             appPL.AppState,
-		interval:        1, // default to every block
-		current:         -1,
+
+		providers: map[string]types.StateProvider{},
+		appState:  &types.AppState{},
+
+		// By default, a snapshot is triggered after each commit.
+		intervalBetweenSnapshots: 1,
+		// This must not be set to 0, otherwise a snapshot will be generated
+		// right after the state has been reload, leading effectively to a
+		// duplicated snapshot.
+		commitsLeftBeforeSnapshot: 1,
```

Comment:
I believe this (previously known as "current") was set to -1 to mean "unset" so on the first call to `OnSnapshotIntervalUpdate` it'll stop us taking a snapshot on block 1 e.g if the interval is set to to 1000 you'd expect snapshots on `1000, 2000, 3000`, but we were seeing snapshots on `1, 1001, 2002`. I have a vague memory of snapshots on block height 1 interacting badly with checkpoint restores.

Anyway, it would be good to just double check that we still do not snapshot on height 1 when `intervalBetweenSnapshots > 1`.

Reply:
Normally, I kept the behavior as before.

Here is the way I understand the previous behavior, assuming `intervalBetweenSnapshots` is set to `1000`.

It's initialized to `-1`. From there 2 paths:
1. a snapshot is applied (state-sync or local, it doesn't matter), once the state is restored, the `commitsLeftBeforeSnapshot` is set to `intervalBetweenSnapshots`. And it decreases (becomes `999`) when Tendermint calls the ABCI method `Commit`. It's not less or equal to `0` so it doesn't snapshot. We don't want to create a snapshot.
2. no snapshot is applied, the `commitsLeftBeforeSnapshot` is left at `-1`. So, when the ABCI method `Commit` is called, it's already below `0`, it get reset and the snapshot is taken. So, I believe it will take a snapshot on the first block.

In our case, setting the `commitsLeftBeforeSnapshot` to the value of `intervalBetweenSnapshots` (instead of `-1`) should yield the same result when a snapshot is applied. When loading from the local storage, there is effectively no difference between setting the value before or after the state restoration. When loading from state-sync, it _shouldn't_ change anything as well as Tendermint shouldn't call the ABCI method `Commit()` on the node when state-sync is enabled. Therefore, it's akin to restoring for the local storage. However, it differs from the previous implementation when nothing is restored, and we are starting from a brand new chain. It would still need to go through the counter to take the snapshot, whereas, in the previous implementation, we were already below `0`, triggering the snapshot on the first call to `Commit()`. Since it didn't feel right, I changed it so it's consistent.